src/irc/mod.rs

```diff
@@ -46,94 +48,105 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
 }
 
 #[derive(Debug, Clone, Default)]
-struct UserNick {
+struct UserNickBuilder {
     nick: Option<String>,
     user: Option<String>,
     real_name: Option<String>,
     password: Option<String>,
 }
 
-impl<S: AsyncRead + AsyncWrite + 'static> IrcUserConnection<S> {
+impl UserNickBuilder {
+    fn to_user_nick(self) -> UserNick {
+        UserNick {
+            nick: self.nick.unwrap(),
+            user: self.user.unwrap(),
+            real_name: self.real_name.unwrap(),
+            password: self.password.unwrap(),
+        }
+    }
+    fn is_complete(&self) -> bool {
+        self.nick.is_some()
+            && self.user.is_some()
+            && self.real_name.is_some()
+            && self.real_name.is_some()
+            && self.password.is_some()
+    }
+}
+
+impl crate::stream_fold::StateUpdate<Result<IrcCommand, io::Error>> for UserNickBuilder {
+    fn state_update(&mut self, new_item: Result<IrcCommand, io::Error>) -> bool {
+        let new_item = match new_item {
+            Ok(good_update) => good_update,
+            Err(_) => return false,
+        };
+
+        match new_item {
+            IrcCommand::Nick { nick } => self.nick = Some(nick),
+            IrcCommand::User { user, real_name } => {
+                self.user = Some(user);
+                self.real_name = Some(real_name);
+            }
+            IrcCommand::Pass { password } => self.password = Some(password),
+            _c => {
+                //debug!(ctx.logger, "Ignore command during login"; "cmd" => c.command());
```

Comment:
add this back

Reply:
Why did you mark as resolved?

Reply:
This one is also fixed on my branch but is not carrying over correctly (?) https://github.com/VanillaBrooks/matrix-ircd/blob/async-await/src/irc/mod.rs#L103-L106. Any idea whats going on here ?